src/Dscp/Crypto/Encrypt.hs

```diff
@@ -125,37 +131,107 @@ authHeader = ""
+cipherBlkSize = blockSize fakeCipher
+  where
+    fakeCipher :: CipherType
+    fakeCipher =
+        error "impossible: cipher constructor is evaluated in `blockSize`"
+
+-------------------------------------------------------------
+-- Encryption/decryption helper datatypes
+-------------------------------------------------------------
+
+-- | Datatype which combines ciphertext with cipher authentication tag.
+data Encrypted ba = Encrypted
+    { eAuthTag    :: !AuthTag
+      -- ^ Authentication tag (to determine whether or not the password is valid)
+    , eIV         :: !(IV CipherType)
+      -- ^ Encryption initial vector (randomly generated during encryption for safety)
+    , eCiphertext :: !ba
+      -- ^ Ciphertext itself
+    } deriving (Eq)
+
+-- | We have to define 'Serialise' instance for 'Encrypted' here,
+-- because we don't export its constructor and field accessors.
+instance FromByteArray ba => Serialise (Encrypted ba) where
+    encode Encrypted {..} =
+        encodeBytes (BA.convert eAuthTag) <>
+        encodeBytes (BA.convert eIV) <>
+        encodeBytes (BA.convert eCiphertext)
+    decode = do
+        eAuthTag <- AuthTag . BA.convert <$> decodeBytes
+        eIV <- maybe (fail "Encrypted: invalid IV size") pure .
+            makeIV =<< decodeBytes
+        eCiphertext <- either fail pure .
+            fromByteArray =<< decodeBytes
+        return Encrypted {..}
+
+instance FromByteArray ba => Buildable (Encrypted ba) where
+    build = build . toBase64 . BSL.toStrict . serialise
+
+instance FromByteArray ba => Show (Encrypted ba) where
+    show = toString . pretty
+
+-------------------------------------------------------------
+-- Encryption/decryption logic
+-------------------------------------------------------------
+
+-- | Derive encryption key from passphrase using PBKDF2
+-- with SHA-512 hash function.
+keyFromPassPhrase :: PassPhrase -> ByteString
+keyFromPassPhrase (PassPhrase pp) = PBKDF2.generate
+    (PBKDF2.prfHMAC SHA512)
+    (PBKDF2.Parameters
+        500         -- Number of hash iterations
+        maxKeySize) -- Encryption key size
+    pp
+    passGenSalt
 
 -- | Prepare an AEAD context from a 'PassPhrase'.
-prepareAEAD :: PassPhrase -> AEAD CipherType
-prepareAEAD (PassPhrase pp) =
+prepareAEAD :: PassPhrase -> IV CipherType -> AEAD CipherType
+prepareAEAD pp iv =
     let impossible err =
-            error $ "encrypt: impossible: " <> Prelude.show err
-        ppHashKey :: ByteString =
-            BA.convert $ hash pp
+            error $ "prepareAEAD: impossible: " <> Prelude.show err
+        ppHashKey =
+            keyFromPassPhrase pp
         cipher :: CipherType =
             onCryptoFailure impossible identity $
             cipherInit ppHashKey
     in onCryptoFailure impossible identity $
-       aeadInit aeadMode cipher initIV
+       aeadInit aeadMode cipher iv
 
 -- | Encrypt given 'ByteArray' with AES.
+-- TODO: use Secure Random from OpenSSL for IV calculation instead.
```

Comment:
Why do you need it?

Reply:
For the same reason why we might need it at all - to make random more _random_ where _randomness_ matter. Do you claim that there's no need for this and using Secure Random(tm) from OpenSSL is simply irrelevant?

Reply:
I doubt that we actually need to use openssl. It's not like it's the only way to generate random data properly.